aptos-move/e2e-move-tests/src/harness.rs

```diff
@@ -797,6 +797,51 @@ macro_rules! assert_abort {
     }};
 }
 
+/// Helper to assert transaction aborts.
+#[macro_export]
+macro_rules! assert_abort_ref {
+    // identity needs to be before pattern (both with and without message),
+    // as if we pass variable - it matches the pattern arm, but value is not used, but overriden.
+    // Opposite order and test_asserts_variable_used / test_asserts_variable_used_with_message tests
+    // would fail
+    ($s:expr, $c:ident $(,)?) => {{
+        claims::assert_matches!(
+            $s,
+            &aptos_types::transaction::TransactionStatus::Keep(
+                aptos_types::transaction::ExecutionStatus::MoveAbort { code, .. }
+            )
+            if code == $c,
+        );
+    }};
+    ($s:expr, $c:pat $(,)?) => {{
+        claims::assert_matches!(
+            $s,
+            &aptos_types::transaction::TransactionStatus::Keep(
+                aptos_types::transaction::ExecutionStatus::MoveAbort { code: $c, .. }
+            )
+        );
+    }};
+    ($s:expr, $c:ident, $($arg:tt)+) => {{
```

Comment:
The `assert_abort_ref` macro has 4 expressions of type `($s: expr, ...) => {..}`. What does that mean?

Reply:
it means, based on what is passed, different variant is instantiated. each of the four branches matches a different thing. you can check unit tests for variety of ways to call it (i.e. with message text and without, etc)

Reply:
huh I didn't know, macro power user level ++